### ISSUE TITLE
TF 2.0.0-alpha0 doesn't build on FreeBSD: error in hwloc

### ISSUE THREAD

**Issue by arrowd** [tensorflow/tensorflow#28113]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): FreeBSD 12.0
- Bazel version (if compiling from source): 0.23.0
- GCC/Compiler version (if compiling from source): Clang 8.0.0

I'n trying to build TensorFlow from source on FreeBSD 12. During build the following problem shows up:

```
ERROR: /usr/home/arr/.cache/bazel/_bazel_arr/5d34786d346adb404a81c63780107a71/external/hwloc/BUILD.bazel:212:1: C++ compilation of rule '@hwloc//:hwloc' failed (Exit 1)
external/hwloc/hwloc/topology-linux.c:39:10: fatal error: 'mntent.h' file not found
#include <mntent.h>
         ^~~~~~~~~~
1 error generated.
```

This is because `topology-linux.c` is being compiled instead of `topology-freebsd.c`. I have zero knowledge in Bazel, so I wasn't able to include on of these files based on the condition.

This should be easy to fix for anyone experienced with Bazel.

**Comment by freedomtan** [comment]
An easy fix is to skip building `hwloc`. With https://github.com/tensorflow/tensorflow/pull/25471 and some minor additional changes (https://github.com/freedomtan/tensorflow/commit/ffd86f605bfdbe6f209c1fa707089d6010f0b719), I can build tensorflow master on FreeBSD.

**Comment by byronyi** [comment]
Ping @gunan; this might not on your radar but in case we have contributors you could offer some advice.

**Comment by arrowd** [comment]
I managed to build TF 2.0.0-alpha with hwloc enabled, however I had to apply a [handful of hacks](http://arrowd.name/tf.patch). All of them a pretty simple, it would be great if someone push them upstream.

**Comment by freedomtan** [comment]
@arrowd clean up your patches and send pull request then. But I think some of the hacks are in #25471, so you may wanna check it first.

**Comment by achandraa** [comment]
@arrowd : Did you get the chance to look into @freedomtan's suggestion ? Is this still an issue. Please let us know. Thanks!

**Comment by arrowd** [comment]
@freedomtan 's branch contains namy changes that I also have. I'll merge them with my patches and create a new pull request.

**Comment by achandraa** [comment]
That's great. Shall be looking forward to it. Thanks!

**Comment by arrowd** [comment]
I got blocked by TF not supporting Bazel 0.25. Is there plans to allow building using new Bazel version?

**Comment by arrowd** [comment]
It is funny that I get both these errors:

>You have bazel 0.23.0 installed.
>Please upgrade your bazel installation to version 0.24.1 or higher to build TensorFlow!

>You have bazel 0.25.0 installed.
>Please downgrade your bazel installation to version 0.24.1 or lower to build TensorFlow!

**Comment by arrowd** [comment]
I need some help with Bazel. How do I implement following changes without braking Linux case?

```
--- third_party/hwloc/BUILD.bazel.orig	2019-04-25 07:56:49 UTC
+++ third_party/hwloc/BUILD.bazel
@@ -36,7 +36,6 @@ template_rule(
         "#undef HWLOC_VERSION_RELEASE": "#define HWLOC_VERSION_RELEASE 3",
         "#undef HWLOC_VERSION_GREEK": "#define HWLOC_VERSION_GREEK \"\"",
         "#undef HWLOC_VERSION": "#define HWLOC_VERSION \"2.0.3\"",
-        "#undef HWLOC_LINUX_SYS": "#define HWLOC_LINUX_SYS 1",
         "#undef hwloc_pid_t": "#define hwloc_pid_t pid_t",
         "#undef hwloc_thread_t": "#define hwloc_thread_t pthread_t",
         "#  undef HWLOC_HAVE_STDINT_H": "#  define HWLOC_HAVE_STDINT_H 1 ",
@@ -86,7 +85,6 @@ _INCLUDE_PRIVATE_HWLOC_AUTOIGEN_CONFIG_H_COMMON_SUBS =
     "#undef HAVE_NL_LANGINFO": "#define HAVE_NL_LANGINFO 1",
     "#undef HAVE_OPENAT": "#define HAVE_OPENAT 1",
     "#undef HAVE_POSIX_MEMALIGN": "#define HAVE_POSIX_MEMALIGN 1",
-    "#undef HAVE_PROGRAM_INVOCATION_NAME": "#define HAVE_PROGRAM_INVOCATION_NAME 1",
     "#undef HAVE_PTHREAD_T": "#define HAVE_PTHREAD_T 1",
     "#undef HAVE_PUTWC": "#define HAVE_PUTWC 1",
     "#undef HAVE_SETLOCALE": "#define HAVE_SETLOCALE 1",
@@ -149,7 +147,6 @@ _INCLUDE_PRIVATE_HWLOC_AUTOIGEN_CONFIG_H_COMMON_SUBS =
     "#undef HWLOC_HAVE_SYSCALL": "#define HWLOC_HAVE_SYSCALL 1",
     "#undef HWLOC_HAVE_X11_KEYSYM": "#define HWLOC_HAVE_X11_KEYSYM 1",
     "#undef HWLOC_HAVE_X86_CPUID": "#define HWLOC_HAVE_X86_CPUID 1",
-    "#undef HWLOC_LINUX_SYS": "#define HWLOC_LINUX_SYS 1",
     "#undef HWLOC_SIZEOF_UNSIGNED_INT": "#define HWLOC_SIZEOF_UNSIGNED_INT 4",
     "#undef HWLOC_SIZEOF_UNSIGNED_LONG": "#define HWLOC_SIZEOF_UNSIGNED_LONG 8",
     "#undef HWLOC_SYM_PREFIX": "#define HWLOC_SYM_PREFIX hwloc_",
@@ -224,13 +221,12 @@ cc_library(
         "hwloc/static-components.h",
         "hwloc/topology.c",
         "hwloc/topology-hardwired.c",
-        "hwloc/topology-linux.c",
+        "hwloc/topology-freebsd.c",
         "hwloc/topology-noos.c",
         "hwloc/topology-synthetic.c",
         "hwloc/topology-xml.c",
         "hwloc/topology-xml-nolibxml.c",
         "hwloc/traversal.c",
-        "include/hwloc/linux.h",
         "include/hwloc/plugins.h",
         "include/hwloc/shmem.h",
         "include/private/autogen/config.h",
```

**Comment by achandraa** [comment]
@arrowd : Can we open a new issue (bug/feature) if it is not related to the issue at hand. We would be happy to help on that and also it would be very easy for us to track. Meanwhile can we close this since it is resolved. Let us know. Thanks!

**Comment by freedomtan** [comment]
@arrowd if you haven't figured out how to do it yet, check bazel's [select](https://docs.bazel.build/versions/master/be/functions.html#select). See TensorFlow's [BUILD.bazel](https://github.com/tensorflow/tensorflow/blob/master/third_party/aws/BUILD.bazel) of AWS SDK for example usage of `select`.

**Comment by achandraa** [comment]
@arrowd :  Let us know if that helps. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by arrowd** [comment]
This is mostlt fixed by https://github.com/tensorflow/tensorflow/pull/28689
